File: frontend/src/utils/xssUtils.js
/**
 * XSS攻击检测与防御工具类
 */

// 常见XSS向量示例
export const xssVectors = [
  {
    name: '基本脚本',
    payload: '<script>alert("XSS")</script>',
    type: 'reflected'
  },
  {
    name: '图片标签事件',
    payload: '<img src="x" onerror="alert(\'XSS\')">',
    type: 'dom'
  },
  {
    name: '链接JavaScript伪协议',
    payload: '<a href="javascript:alert(\'XSS\')">点击我</a>',
    type: 'dom'
  },
  {
    name: 'CSS表达式',
    payload: '<div style="background-image: url(javascript:alert(\'XSS\'))"></div>',
    type: 'reflected'
  },
  {
    name: '内联事件监听器',
    payload: '<body onload="alert(\'XSS\')"></body>',
    type: 'dom'
  }
];

// 检测字符串中的潜在XSS载荷
export const detectXssPayload = (input) => {
  if (!input || typeof input !== 'string') return false;
  
  // 定义可能触发XSS的模式
  const patterns = [
    /<script\b[^>]*>(.*?)<\/script>/i,
    /javascript\s*:/i,
    /on\w+\s*=/i,
    /<\s*img[^>]+onerror\s*=/i,
    /<\s*iframe/i,
    /<\s*embed/i,
    /<\s*object/i,
    /data\s*:/i,
    /document\.cookie/i,
    /document\.location/i,
    /eval\s*\(/i,
    /document\.write/i,
    /innerHTML/i,
    /fromCharCode/i
  ];
  
  // 检查是否匹配任何一种模式
  return patterns.some(pattern => pattern.test(input));
};

// 分析XSS有效载荷的危险等级
export const analyzeXssSeverity = (payload) => {
  if (!payload) return 'low';
  
  const severePatterns = [
    /document\.cookie/i,
    /localStorage/i, 
    /sessionStorage/i,
    /\.src\s*=/i,
    /xhr\.open/i,
    /fetch\s*\(/i,
    /new\s+XMLHttpRequest/i
  ];
  
  const mediumPatterns = [
    /<script/i,
    /eval\s*\(/i,
    /document\.write/i,
    /innerHTML/i
  ];
  
  if (severePatterns.some(pattern => pattern.test(payload))) {
    return 'high';
  } else if (mediumPatterns.some(pattern => pattern.test(payload))) {
    return 'medium';
  } else {
    return 'low';
  }
};

// 解析CSP策略字符串
export const parseCspPolicy = (policyString) => {
  if (!policyString) return {};
  
  const result = {};
  const directives = policyString.split(';').map(dir => dir.trim());
  
  directives.forEach(directive => {
    if (directive) {
      const [key, ...values] = directive.split(/\s+/);
      result[key] = values;
    }
  });
  
  return result;
}; 

File: backend/src/main/java/com/xssdefense/system/util/XssUtils.java
package com.xssdefense.system.util;

import java.util.regex.Pattern;

/**
 * XSS攻击检测和防御工具类
 */
public class XssUtils {

    /**
     * 检测字符串中的XSS载荷
     */
    public static boolean detectXssPayload(String input) {
        if (input == null || input.isEmpty()) {
            return false;
        }
        
        // 定义可能触发XSS的模式
        String[] patterns = {
            "<script[^>]*>.*?</script>",
            "javascript\\s*:",
            "on\\w+\\s*=",
            "<\\s*img[^>]+onerror\\s*=",
            "<\\s*iframe",
            "<\\s*embed",
            "<\\s*object",
            "data\\s*:",
            "document\\.cookie",
            "document\\.location",
            "eval\\s*\\(",
            "document\\.write",
            "innerHTML",
            "fromCharCode"
        };
        
        // 检查是否匹配任何一种模式
        for (String pattern : patterns) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE | Pattern.DOTALL).matcher(input).find()) {
                return true;
            }
        }
        
        return false;
    }
    
    /**
     * 分析XSS载荷的危险等级
     */
    public static String analyzeXssSeverity(String payload) {
        if (payload == null || payload.isEmpty()) {
            return "low";
        }
        
        String[] severePatterns = {
            "document\\.cookie",
            "localStorage",
            "sessionStorage",
            "\\.src\\s*=",
            "xhr\\.open",
            "fetch\\s*\\(",
            "new\\s+XMLHttpRequest"
        };
        
        String[] mediumPatterns = {
            "<script",
            "eval\\s*\\(",
            "document\\.write",
            "innerHTML"
        };
        
        // 检查是否匹配高危模式
        for (String pattern : severePatterns) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(payload).find()) {
                return "high";
            }
        }
        
        // 检查是否匹配中危模式
        for (String pattern : mediumPatterns) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(payload).find()) {
                return "medium";
            }
        }
        
        return "low";
    }
    
    /**
     * 净化可能包含XSS的HTML内容
     */
    public static String sanitizeHtml(String html) {
        if (html == null || html.isEmpty()) {
            return "";
        }
        
        return html
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;")
            .replace("`", "&#96;")
            .replace("(", "&#40;")
            .replace(")", "&#41;")
            .replaceAll("(?i)javascript:", "")
            .replaceAll("(?i)on\\w+=", "");
    }
} 